### PR TITLE
Improve working with message lists and calendars in Outlook

### LINKED ISSUE #4971: NVDA does not report status of messages in Outlook 2010
**Reported by andrewd on 2015-03-10 21:52**
NVDA reports when a message is unread.  It does not report when a message has been read, replied to or forwarded.

### LINKED ISSUE #6911: NVDA not inform about if an e-mail in Outlook 2013 is answered or forwarded
Hi everyone

If I want to read messages in Outlook 2013 and use the arrow keys to scroll up and down in my folders (Inbox, Sent Mail and more), unlike JAWS, NVDA does not announce if an e-mail is replied / forwarded.
Every time that NVDA comes to an e-mail that is either replied / forwarded, Nvda only announce that the e-mail have an attached file. The e-mails not has an attached file.

Is it an error in NVDA?
If yes, I will be glad for a correction.

Thanks in advance.

Regards,
Anders

### LINKED ISSUE #8028: NVDA reports "has attachment" for messages without visible attachments in Outlook
### Steps to reproduce:
1. In outlook, send yourself a message containing a picture in the message body.
2. List the message in the Inbox

### Expected behavior:
The picture is part of the message body, so NVDA shouldn't announce "has attachments".

### Actual behavior:
NVDA announces has attachment. However, when you press enter on the message to open it, than shift+tab to go to the list of attachments, there simply is no such list.

### Additional information
* The attachments cell for this grid item is named "No attachments"
* This issue is caused by the Outlook object model not only exposing explicitly added attachments, but also implicitly added ones (i.e. in a message's body).

### Proposed fix
Remove the attachment checking altogether and rely on the UIA information. For this, the children cache request we use to fetch the child nodes of a grid row should no longer only list text nodes. Note that this also allows us to rely on UIA information for the importance status. I think relying on UIA here is preferred due to the following reasons:

1. It fixes the attachments bug as pointed out above
2. It respects different view setups for outlook. Imagine a case where one edits his view to make it expose the attachments status as text rather than as an icon. This would lead to double speaking of information.

It is not advised to rely on UIA only for the flag status, as unflagged message grid rows still expose a child node named "unflagged". This is redudant information, but can easily be worked around by filtering on the UIA Class name "FlagIcon".

### System configuration:
NVDA version: 2017.4
NVDA Installed
Windows version: Windows 10 fall creators update
Outlook version: 16.0.9001.2102

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #6911
Fixes #8028
Reverts #6219

### Summary of the issue:
1. In the Outlook messages list, replied to or forwarded status is not announced by NVDA.
2. In outlook's calendar view, the month view is overly verbose, announcing redundant information (e.g. 00.00AM to next day 00.00AM
3. When a message has an associated draft, this is not reported.

### Description of how this pull request fixes the issue:
1. This pr gets the announcement of replied/forwarded state from the UIA value of rows in the Outlook messages list. NVDA will now also announce meeting requests based on this information, as well as appointment cancellations.

    The initial implementation of this pr relied on the Outlook object model. However, this caused security warnings on some systems with a tight security policy.

2. When an appointment or time slot is now exactly one day (86400 seconds), and starts at 00.00, the time slot is announced as `26 february 2018 (entire day)` instead of `26 februari 2018 00.00 to 27 februari 2018 00.00`. While looking at this code, I noticed that Microsoft strongly discourages the use of kernel32 GetTimeFormat and GetDateFormat functions, so I changed this implementation to use the *Ex alternatives for these functions.

3. The associated draft column in Outlook 2016 is no longer silenced. After implementing #6219, it seems Microsoft decided to present an empty text node when there is no associated draft, rather than spamming everyone with the "no associated draft" message. Therefore when this particular pr is merged, everyone who gets "no associated draft" for almost every message is advised to update Outlook 2016 to the most recent build.

### Testing performed:
1. Made sure that the current implementation doesn't cause security warnings when setting outlook's program access settings to the most secure option available.
2. Tested replied/forwarded messages in Outlook 2016 and 2013, the state was reported correctly. The replied state did even report for messages I replied to using Thunderbird. Also tested this in Outlook 2010, and as Outlook 2010 doesn't expose this information using UIA, we fall back to the object model for the unread state.
3. Tested the announcement of a whole day appointment in the calendar

### Known issues with pull request:
1. The replied or forwarded state is localized by Outlook itself, not by NVDA. This shouldn't be a problem though, as the headers are also localized by Outlook itself.

2. In theory, Outlook could expose every piece of relevant or irrelevant information using the UIA value. I've only seen the following strings until now:
    * Message: usually the first word in the value of a message
    * Contact: usually the first word in the value of a contact when the contacts are displayed in a list view
    * Appointment: usually the first word in the value of a calendar item when the calendar is displayed in a list view
    * Read/Unread: usually the last word in the value of a message, contact or appointment
    * Forwarded/Replied: usually the second word in a string containing three words, e.g. Message Replied Read

    I'm quite sure JAWS also relies on the value here, as replied/forwarded state is localized using Outlook, not using JAWS. The wild should make clear whether parsing the value this way is too non-restrictive.

3. The replied to all state is not reported as it is not exposed using the object's value.

4. When the root of a conversation in conversation view has focus, NVDA reads the headers of every single message in the conversation. There is currently no suitable workaround for this. Note that this was already the case before this pr. This issue is caused by a bug in UIA, where the tree scope is ignored when providing a custom tree filter for a cache request.

### Change log entry:
* New features
    + In Microsoft Outlook message lists, NVDA now reports if a message has been replied to or forwarded. (#6911)

* Changes
    + NVDA is now less verbose when an appointment or time slot in the Outlook calendar covers an entire day. (#7949)

### CODE REVIEW

**Review by michaelDCurran — DISMISSED** [review]

**Comment by leonardder** [comment]
Blocked by #7953

**Comment by leonardder** [comment]
I had to change this into a completely different implementation due to #7953. @MichaelDCurran. It would be great if you could review this as soon as possible to fix #7953.

File: source/appModules/outlook.py
Comment (michaelDCurran) [inline source/appModules/outlook.py]: Can we be sure that the first part (message, contact etc) is never more than one word in all languages?

Comment (leonardder) [reply source/appModules/outlook.py]: Good question, which also applies to read/unread I belief.

Though JAWS uses these translated strings, I"m pretty sure they do some string matching on it, as the read state is ignored when Both Outlook and JAWS are set to Dutch, but the state is preserved when Outlook is Dutch and JAWS is English. So that's definitely not a clean solution.

I belief both Message and Unread are part of the NVDA translation, so it should be doable to check whether these words contain spaces in other languages without intervention of every single translator.

**Comment by leonardder** [comment]
@michaeldcurran: I made the scope of this pr somewhat broader and also addressed some other concerns there were with the current way of dealing with Outlook. The initial message of the pr covers all relevant information.

File: source/appModules/outlook.py
Comment (dkager) [inline source/appModules/outlook.py]: So there is no such class for a message/email?

File: source/appModules/outlook.py
Comment (dkager) [inline source/appModules/outlook.py]: This also catches events that start at noon on one day and end at noon on the next. Do we want to call these all day events or should we check for the start and end times being midnight as well?

File: source/appModules/outlook.py
Comment (dkager) [inline source/appModules/outlook.py]: Outlook calls them "All day event", not "whole day".

File: source/appModules/outlook.py
Comment (dkager) [inline source/appModules/outlook.py]: Again, "all day" may be better here.

File: source/appModules/outlook.py
Comment (dkager) [inline source/appModules/outlook.py]: Yet the example you give above is "Meeting request". Is this not regular? If so, 2 words should be filtered out.

File: source/appModules/outlook.py
Comment (dkager) [inline source/appModules/outlook.py]: Does this handle the "DraftFlagField" case?

File: source/appModules/outlook.py
Comment (leonardder) [reply source/appModules/outlook.py]: Yes, that's the note one. I added a comment that clarifies this.

File: source/appModules/outlook.py
Comment (leonardder) [reply source/appModules/outlook.py]: Good point, I forgot to mention this in the initial pr description, I surely should have done that. It is still open to discussion.

MS no longer exposes "no associated draft" as a text node by default. Instead, the text node is empty. Furthermore, the "no associated draft" patch meant that it would be never announced when a message has an associated draft. Therefore, I propose people who have the "no associated draft" issue updating their outlook to fix this, rather than keeping the old patch. CC @bramd

**Comment by derekriemer** [comment]
what about saying "all day" instead of "entire day"

**Comment by leonardder** [comment]
[**@derekriemer**](https://github.com/derekriemer) commented on [5 mrt. 2018 15:53 CET](https://github.com/nvaccess/nvda/pull/7949#issuecomment-370443908 "2018-03-05T14:53:20Z - Replied by Github Reply Comments"):
> what about saying "all day" instead of "entire day"

It is now saying "all day" as per the last commits from today morning.

File: source/appModules/outlook.py
Comment (dkager) [reply source/appModules/outlook.py]: I concur.

**Comment by leonardder** [comment]
@MichaelDcurran, this pr has been overhauled and is ready for another review I think. I'm afraid there's no way to work around relying on the localized contents of the UIA value.

**Review by michaelDCurran — APPROVED** [review]

**Comment by josephsl** [comment]
Hi,

I wonder if this is the right PR for this: in Outlook 2016's Inbox view, when I perform read current line (caret) command with no messages shown, I get:

**Comment by josephsl** [comment]
Thanks.

**Comment by josephsl** [comment]
Hi,

Er, I don't know how the traceback got cut off...

{{{
ERROR - scriptHandler.executeScript (22:15:33.871):
error executing script: <bound method GlobalCommands.script_reportCurrentLine of <globalCommands.GlobalCommands object at 0x0584E7D0>> with gesture u'NVDA+up arrow'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 187, in executeScript
  File "globalCommands.pyc", line 142, in script_reportCurrentLine
  File "textInfos\offsets.pyc", line 348, in expand
  File "textInfos\offsets.pyc", line 330, in _getUnitOffsets
  File "textInfos\offsets.pyc", line 249, in _getLineOffsets
  File "NVDAObjects\__init__.pyc", line 42, in _getStoryText
  File "baseObject.pyc", line 34, in __get__
  File "baseObject.pyc", line 115, in _getPropertyViaCache
  File "NVDAObjects\__init__.pyc", line 1027, in _get_basicText
  File "baseObject.pyc", line 34, in __get__
  File "baseObject.pyc", line 115, in _getPropertyViaCache
  File "appModules\outlook.pyc", line 451, in _get_name
  File "comtypes\__init__.pyc", line 279, in __getattr__
ValueError: NULL COM pointer access
}}}

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [15 mrt. 2018 06:18 CET](https://github.com/nvaccess/nvda/pull/7949#issuecomment-373263739 "2018-03-15T05:18:20Z - Replied by Github Reply Comments"):
> I wonder if this is the right PR for this.

Looks like it is. It turns out it wasn't a good idea to remove the tree filter from the cache request that forced the fetched children to be text controls. I now added a new tree filter that requires the children to be content children. After that, I couldn't reproduce the issue any more.

I also noticed another bug in the code that tried to access the object model for flag information when there was no selection. That bug was actually obfuscated by the treeFilter one. We are now ignoring flags when there is no object model access, just as we did before this pr.

**Comment by leonardder** [comment]
Currently blocked by #8129. When getting the name for message list items, a NULL COM pointer access error is raised because the custom UIA query fails. This is a bit strange, since until now, it fails only on two systems. @bramd has a system where it fails and another system where it works correctly, same windows and office builds.

Note that such an error has been reported by @josephsl in https://github.com/nvaccess/nvda/pull/7949#issuecomment-373263880, the fix for that has yet not been committed to Next. However, testing by @bramd tells that the fix for https://github.com/nvaccess/nvda/pull/7949#issuecomment-373263880 does not aplly to this particular case of the null com pointer access error.

I guess we can do an if check on cachedChildren, but I'd like to no why getting cached children would reveal a null com pointer in the first place. @michaeldcurran: Have you seen these cases earlier?

**Comment by michaelDCurran** [comment]
I think I have seen a NULL COM pointer for cached children before. But I 
can't remember exactly where.
We could argue it is a Microsoft bug, but in the end, we will still have 
to  handle it if  it is happening in the wild.

**Comment by leonardder** [comment]
@michaeldcurran: Could you please review the last few commits?

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**
